src/gfn/utils/prob_calculations.py

```diff
@@ -0,0 +1,232 @@
+    else:
+        if trajectories.estimator_outputs is not None and not recalculate_all_logprobs:
+            estimator_outputs = trajectories.estimator_outputs[action_mask]
+        else:
+            masked_cond = None
+            if trajectories.conditioning is not None:
+                cond_dim = (-1,) * len(trajectories.conditioning.shape)
+                traj_len = trajectories.states.tensor.shape[0]
+                masked_cond = trajectories.conditioning.unsqueeze(0).expand(
+                    (traj_len,) + cond_dim
+                )[state_mask]
+
+            estimator_outputs = check_cond_forward(pf, "pf", valid_states, masked_cond)
+
+        # Calculates the log PF of the actions sampled off policy.
+        valid_log_pf_actions = pf.to_probability_distribution(
+            valid_states, estimator_outputs
+        ).log_prob(
+            valid_actions.tensor
+        )  # Using the actions sampled off-policy.
+
+        log_pf_trajectories = torch.full_like(
+            trajectories.actions.tensor[..., 0],
+            fill_value=fill_value,
+            dtype=torch.float,
+        )
+        log_pf_trajectories[action_mask] = valid_log_pf_actions
+
+    return log_pf_trajectories
+
+
+def get_traj_pbs(
+    pb: GFNModule, trajectories: Trajectories, fill_value: float = 0.0
+) -> torch.Tensor:
+    # Note the different mask for valid states and actions compared to the pf case.
+    state_mask = (
+        ~trajectories.states.is_sink_state & ~trajectories.states.is_initial_state
+    )
+    action_mask = ~trajectories.actions.is_dummy & ~trajectories.actions.is_exit
+
+    valid_states = trajectories.states[state_mask]
+    valid_actions = trajectories.actions[action_mask]
+
+    if valid_states.batch_shape != tuple(valid_actions.batch_shape):
+        raise AssertionError("Something wrong happening with log_pf evaluations")
+
+    # Using all non-initial states, calculate the backward policy, and the logprobs
+    # of those actions.
+    masked_cond = None
+    if trajectories.conditioning is not None:
+        # We need to index the conditioning vector to broadcast over the states.
+        cond_dim = (-1,) * len(trajectories.conditioning.shape)
+        traj_len = trajectories.states.tensor.shape[0]
+        masked_cond = trajectories.conditioning.unsqueeze(0).expand(
+            (traj_len,) + cond_dim
+        )[state_mask]
+
+    estimator_outputs = check_cond_forward(pb, "pb", valid_states, masked_cond)
+
+    valid_log_pb_actions = pb.to_probability_distribution(
+        valid_states, estimator_outputs
+    ).log_prob(valid_actions.tensor)
+
+    log_pb_trajectories = torch.full_like(
+        trajectories.actions.tensor[..., 0],
+        fill_value=fill_value,
+        dtype=torch.float,
+    )
+    log_pb_trajectories[action_mask] = valid_log_pb_actions
+
+    return log_pb_trajectories
+
+
+########################
+##### Transitions  #####
+########################
+
+
+def get_trans_pfs_and_pbs(
```

Comment:
sorry to be picky about this but I think we should avoid calling this `trans`. Maybe better to stick with `transitions`.

Reply:
Alright!